### ISSUE TITLE
Generic Commit message pattern

### ISSUE THREAD

**Issue by Jmclerck** [jzaefferer/commitplease#87]
I have been trying to use the, admittedly, experimental regex patterns described in the README and I'm struggling to make it do what I want. We have both a very specific first commit line style but a looser long description in the company I work for. I am trying to use the existing patterns to check a first commit line with a style similar to this:

`[XX-01]: Initial commit`

`XX` in this case would be an alphabetical project code. I have tried using the ticket patterns but I have been unable to fathom out how to validate this style with the available options. Would you be able to provide assistance or be open to adding a more generic `firstlinePattern`?

Thanks!

**Comment by all3fox** [comment]
It looks like you want a "dynamic" element `[XX-01]` where both `XX` and `01` keep changing and the "brackets and minus" structure (`[..-..]`) remains unchanged, right?

If you do not have a crazy amount of projects, a quick fix is to add this into `package.json`:

```
{
  "commitplease": {
    "style": "jquery",
    "component": true,
    "components": ["[AB-42]"]
  }
}
```

If that does not work for you or you do not feel like adding it into 42+ projects, let me know

**Comment by Jmclerck** [comment]
Hi @all3fox thanks for your swift response as usual 😸 

That is pretty much what I am looking for. At the moment we are only using `commitplease` on one project so at this time the `XX` would be fixed, although it would need to be configured in the future. 

The digits after the minus change frequently though as they are references to tickets so it would need to an expression like `\d+`. That is why I think another type of expression is going to be the simplest way forward?

**Comment by all3fox** [comment]
Hi, more questions then:

1. Why would you need a different `XX` in the commit history of the project (since you know that you are looking at the history of that project, you know what `XX` has to be). Are you building some "aggregated" project out of many smaller ones?
1. The integer part: is it a project's version or an issue number? If that is an issue number, why not stick with the `closes #123`, `fixes #123` and so on at the end of the message?

I am ready to try and implement this over the weekend, it is just I would like to understand it first.

**Comment by Jmclerck** [comment]
Hi,

1. We have an internal ticket tracking system that is separate to our source control systems, so in order to cross reference commits to the ticket system we always prefix the commit messages with the ticket number which is always in the format `XX-\d+`. We wouldn't have a different `XX` in the same commit history, but the `XX` would change between projects so it would need to be configurable at some point, although I realise that the component suggestion you made would cover this part at least.

2. The integer part is as you say an issue/ticket number, but they are always prefixed with the project code, again the `XX` part. I don't really have any issue with using the `closes/fixes` at the end but this is an internal convention that all other projects use and we need to adopt it, that isn't something that I can change.

I realise this sounds like a strange system but it is what I have to work with, I am hoping `commitplease` can make it easier for new people coming into it 😄

**Comment by jzaefferer** [comment]
Could we extend the `components` check to accept a regex? If its a string, check for exact match, if its a regex, see if it passes...

**Comment by Jmclerck** [comment]
@all3fox + @jzaefferer sounds like it would be fine for my purposes, I'll be happy to test for you and provide feedback. Thanks again!

**Comment by all3fox** [comment]
@Jmclerck please try out the feature by:

`npm install git+https://github.com/jzaefferer/commitplease.git#component-as-regexp`

and then adjusting `package.json` to use a regexp:

```
{
  "commitplease": {
    "style": "jquery",
    "component": true,
    "components": ["^\\[\\w+-\\d+\\]"]
  }
}
```

**Comment by Jmclerck** [comment]
Hi @all3fox, I have just tried out the new branch and it suits my purposes perfectly. I am happy with the modification to the behaviour. I'll leave it to up you to you guys to merge into the project but if there is any more testing to be done, please let me know and thanks again for all your hard work 👍

**Comment by all3fox** [comment]
This change landed on npm as `2.7.5` (or `latest`). Note that there was a last-minute small change: every `component` from the list of `components` gets surrounded by `^component$`, so you can drop the `^` from my previous comment:
```
{
  "commitplease": {
    "style": "jquery",
    "component": true,
    "components": ["\\[\\w+-\\d+\\]"]
  }
}
```
Feel free to come back with more issues/feature requests!
